src/cpp/ext/gsm/metadata_exchange.cc

```diff
@@ -0,0 +1,344 @@
+  if (it == attributes.end()) {
+    return "unknown";
+  }
+  const auto* string_value = absl::get_if<std::string>(&it->second);
+  if (string_value == nullptr) {
+    return "unknown";
+  }
+  return *string_value;
+}
+
+absl::string_view GetStringValueFromUpbStruct(google_protobuf_Struct* struct_pb,
+                                              absl::string_view key,
+                                              upb_Arena* arena) {
+  if (struct_pb == nullptr) {
+    return "unknown";
+  }
+  google_protobuf_Value* value_pb = google_protobuf_Value_new(arena);
+  bool present = google_protobuf_Struct_fields_get(
+      struct_pb, AbslStrToUpbStr(key), &value_pb);
+  if (present) {
+    if (google_protobuf_Value_has_string_value(value_pb)) {
+      return UpbStrToAbslStr(google_protobuf_Value_string_value(value_pb));
+    }
+  }
+  return "unknown";
+}
+
+class LocalLabelsIterable : public LabelsIterable {
+ public:
+  explicit LocalLabelsIterable(
+      const std::vector<std::pair<absl::string_view, std::string>>& labels)
+      : labels_(labels) {}
+
+  absl::optional<std::pair<absl::string_view, absl::string_view>> Next()
+      override {
+    if (pos_ >= labels_.size()) {
+      return absl::nullopt;
+    }
+    return labels_[pos_++];
+  }
+
+  size_t Size() const override { return labels_.size(); }
+
+  void ResetIteratorPosition() override { pos_ = 0; }
+
+ private:
+  size_t pos_ = 0;
+  const std::vector<std::pair<absl::string_view, std::string>>& labels_;
+};
+
+class PeerLabelsIterable : public LabelsIterable {
+ public:
+  explicit PeerLabelsIterable(grpc_core::Slice remote_metadata)
+      : metadata_(std::move(remote_metadata)) {}
+
+  absl::optional<std::pair<absl::string_view, absl::string_view>> Next()
+      override {
+    auto& struct_pb = GetDecodedMetadata();
+    if (struct_pb.struct_pb == nullptr) {
+      return absl::nullopt;
+    }
+    if (++pos_ == 1) {
+      return std::make_pair(kPeerTypeAttribute,
+                            GetStringValueFromUpbStruct(
+                                struct_pb.struct_pb, kMetadataExchangeTypeKey,
+                                struct_pb.arena.ptr()));
+    }
+    // Only handle GKE type for now.
+    switch (type_) {
+      case GcpResourceType::kGke:
+        if (pos_ - 2 >= kGkeAttributeList.size()) {
+          return absl::nullopt;
+        }
+        return std::make_pair(
+            kGkeAttributeList[pos_ - 2].otel_attribute,
+            GetStringValueFromUpbStruct(
+                struct_pb.struct_pb,
+                kGkeAttributeList[pos_ - 2].metadata_attribute,
+                struct_pb.arena.ptr()));
+      case GcpResourceType::kUnknown:
+        return absl::nullopt;
+    }
+  }
+
+  size_t Size() const override {
+    auto& struct_pb = GetDecodedMetadata();
+    if (struct_pb.struct_pb == nullptr) {
+      return 0;
+    }
+    if (type_ != GcpResourceType::kGke) {
+      return 1;
+    }
+    return 8;
```

Comment:
Please change this to use `kGkeAttributeList.size() + 1`.

Reply:
ofcourse, that's what we wanted to do in the first place!